// nativescript/lib/common/appbuilder/project-constants.js
"use strict";
var ProjectConstants = (function () {
    function ProjectConstants() {
        this.PROJECT_FILE = ".abproject";
        this.PROJECT_IGNORE_FILE = ".abignore";
        this.DEBUG_CONFIGURATION_NAME = "debug";
        this.DEBUG_PROJECT_FILE_NAME = ".debug.abproject";
        this.RELEASE_CONFIGURATION_NAME = "release";
        this.RELEASE_PROJECT_FILE_NAME = ".release.abproject";
        this.CORE_PLUGINS_PROPERTY_NAME = "CorePlugins";
        this.CORDOVA_PLUGIN_VARIABLES_PROPERTY_NAME = "CordovaPluginVariables";
        this.APPIDENTIFIER_PROPERTY_NAME = "AppIdentifier";
        this.EXPERIMENTAL_TAG = "Experimental";
        this.NATIVESCRIPT_APP_DIR_NAME = "app";
        this.IMAGE_DEFINITIONS_FILE_NAME = 'image-definitions.json';
        this.PACKAGE_JSON_NAME = "package.json";
        this.ADDITIONAL_FILE_DISPOSITION = "AdditionalFile";
        this.ADDITIONAL_FILES_DIRECTORY = ".ab";
        this.APPBUILDER_PROJECT_PLATFORMS_NAMES = {
            android: "Android",
            ios: "iOS",
            wp8: "WP8"
        };
        this.IONIC_PROJECT_PLATFORMS_NAMES = {
            android: "android",
            ios: "ios",
            wp8: "wp8"
        };
    }
    return ProjectConstants;
}());
exports.ProjectConstants = ProjectConstants;
$injector.register("projectConstants", ProjectConstants);
